---
title: 'BG3 vs S2 analysis'
author: 'Patrycja Rosa'
date: '`r sub("^0", "", format(Sys.time(), "%d %B %Y"))`'
output: html_document
---
> Some introductory text here 


```{r, echo=FALSE, message=FALSE}
library(edgeR)
library(umap)
library(ggplot2)
library(limma)
library(VennDiagram)
library(pheatmap)
library(RColorBrewer)
```

```{r, echo=FALSE}
# for plotting, take the genes with abs(log2 fold change) > 2 and adjusted p-value < 0.05
logfc = 2
alpha = 0.05 

#read counts file 
data <- read.delim('P:/Desktop/work/rna-seq-bg3/all.tsv')
rownames(data) <- data[,1]
data <- data[,-1]
data_cpm <- cpm(data)

#read model statistics value table 
statistics <- read.table('P:/Desktop/work/rna-seq-bg3/ML.DmBG3.c2-vs-S2.DRSC.diffexp.tsv')

#differential expressed genes 
diffexp <- statistics[which(abs(statistics$log2FoldChange)>2 & statistics$padj < 0.05),]
diff_genes <- rownames(diffexp)
diff_counts <- data_cpm[which(rownames(data_cpm) %in% rownames(diffexp)),]



#pheatmap 
pheatmap(mat = log10(diff_counts+1),
         col = colorRampPalette(brewer.pal(8, "PuBuGn"))(20),
         cluster_cols = FALSE,
         show_rownames = FALSE)

#edgeR analysis
samples <- c("Ml_DmBG3.c2", "Ml_DmBG3.c2", "S2.DRSC", "S2.DRSC")
samples <- cbind(samples, colnames(data))
genes <- rownames(data)
rownames(samples) <- colnames(data)

#make dge object for further analysis
dge_object <- DGEList(counts = data, samples = samples, genes = genes)

#filtering by variance
keep <- apply(dge_object$counts,1,var)>0

#sum(keep)/length(keep)
data_filtered <- dge_object[keep, , keep.lib.sizes=TRUE]

#normalization
data_norm<-calcNormFactors(data_filtered)

#model matrix
design <- model.matrix(~0 +data_norm$samples$samples)
rownames(design) <- data_norm$samples$V2

#estimate dispertion
y <- estimateDisp(data_norm, design, robust = TRUE)

#fit the model into data
fit <- glmQLFit(y, design)

#qlf test
qlf_data <- glmQLFTest(fit, contrast = c(1,-1))

#multiple testing correction method
FDR.data <- p.adjust(qlf_data$table$PValue, method = "BH")

#add p-adjusted value into summary table
qlf_data$table$adjPValue <- FDR.data


#choose genes with abs(logfoldchange) >2 and adjusted p-value<0.05
thresh_data <- qlf_data[which(abs(qlf_data$table$logFC)>logfc & qlf_data$table$adjPValue<alpha),]

#find genes from edgeR in deseq differential expressed genes
rownames(thresh_data$table) -> edger_genes
com_genes <- edger_genes[which(diff_genes %in% edger_genes)]

```